yt_dlp/extractor/sheeta.py

```diff
@@ -0,0 +1,836 @@
+                return self._auth_info.get('auth_token')
+            except Exception as e:
+                raise ExtractorError('Unable to refresh token due to unknown reasons') from e
+
+        return None
+
+    def _refresh_token(self):
+        if not (refresh_func := self._auth_info.get('refresh_func')):
+            return False
+
+        res = self._download_json(
+            **refresh_func(self._auth_info), expected_status=(400, 403, 404),
+            note='Refreshing token', errnote='Unable to refresh token')
+        if error := traverse_obj(
+                res, ('error', 'message', {lambda x: base64.b64decode(x).decode()}), ('error', 'message')):
+            self.report_warning(f'Unable to refresh token: {error!r}')
+        elif token := traverse_obj(res, ('data', 'access_token', {str})):
+            # niconico
+            self._auth_info = {'auth_token': f'Bearer {token}'}
+            return True
+        elif token := traverse_obj(res, ('access_token', {str})):
+            # auth0
+            self._auth_info = {'auth_token': f'Bearer {token}'}
+            if refresh_token := traverse_obj(res, ('refresh_token', {str})):
+                self._auth_info = {'refresh_token': refresh_token}
+                self.cache.store(
+                    self._NETRC_MACHINE, self._auth_info['cache_key'], {self._auth_info['cache_name']: refresh_token})
+                return True
+            self.report_warning('Unable to find new refresh_token')
+        else:
+            self.report_warning('Unable to refresh token')
+
+        return False
+
+    def _login(self):
+        social_login_providers = traverse_obj(self._call_api(
+            f'fanclub_groups/{self._FANCLUB_GROUP_ID}/login', None),
+            ('data', 'fanclub_group', 'fanclub_social_login_providers', ..., {dict})) or []
+        self.write_debug(f'social_login_providers = {social_login_providers!r}')
+
+        for provider in social_login_providers:
+            provider_name = traverse_obj(provider, ('social_login_provider', 'provider_name', {str}))
+            if provider_name == 'ニコニコ':
+                redirect_url = update_url_query(provider['url'], {
+                    'client_id': 'FCS{:05d}'.format(provider['id']),
+                    'redirect_uri': f'https://{self._DOMAIN}/login',
+                })
+                refresh_url = f'{self._API_BASE_URL}/fanclub_groups/{self._FANCLUB_GROUP_ID}/auth/refresh'
+                return self._niconico_sns_login(redirect_url, refresh_url)
+            else:
+                raise ExtractorError(f'Unsupported social login provider: {provider_name}')
+
+        return self._auth0_login()
+
+    def _niconico_sns_login(self, redirect_url, refresh_url):
+        self._auth_info = {'login_method': 'any'}
+        mail_tel, password = self._get_login_info()
+        if not mail_tel:
+            return
+
+        cache_key = hashlib.sha1(f'{self._DOMAIN}:{mail_tel}:{password}'.encode()).hexdigest()
```

Comment:
Why are we caching this to disk? From what I can tell, the data we are caching doesn't depend on any responses from the site, we are just caching user and password to disk? And it will only be loaded if the user is already passing their username and password?

Reply:
Here I'm caching tokens for successfully logged-in users.

Almost all sheeta-based websites (except https://nicochannel.jp/) don't support cookie-based authorization, so login credentials are required. If a user runs yt-dlp multiple times without cached tokens, time will be wasted submitting login information.

This is emulating a web browser. The JS code stores tokens in Local Storage, so users don't have to type password every time they enter sheeta sites.

To keep each user's token distinct from anyone else's on each website, I'm using a hash algorithm to make the cache key.

Reply:
`cache_key` just seems unnecessary to me, and there is definitely no reason to be hashing the password. Using `self._NETRC_MACHINE` as the `section` should do the work of distinguishing between different sites, and the `key` can just be the username

Reply:
or if for some reason you do need `_DOMAIN` instead of simply `self._NETRC_MACHINE`, the password can still be omitted from the key

Reply:
Hmm, this IE supports multiple sites, but `self._NETRC_MACHINE` is a constant string. If a user uses the same username (e-mail address) on different sites, the cache will be overwritten. So, I think it's not OK to use `self._NETRC_MACHINE` as the `section`.

About hashing password: Now I think that's totally unnecessary. If a cached token is invalidated for some reason, our code will drop it and try getting a new one. We shouldn't care if the password is changed or not.

I still need a `cache_key` since I need update cached tokens outside of the login procedure.

So:

```python
cache_key = base64.b64decode(f'{self._DOMAIN}:{mail_tel}'.encode()).decode()

```

Your opinion?

Reply:
> Hmm, this IE supports multiple sites, but self._NETRC_MACHINE is a constant string.

We should not be subclassing these other sites' extractors from a concrete IE like `SheetaEmbedIE`. There should be a `SheetaBaseIE` from which they are subclassed. This will also simplify it so we can set the `_NETRC_MACHINE` in the subclasses and it will be unique per site

Reply:
Well no. There is no need to subclass other sites' extractors. This extractor supports all sites that are using the Sheeta system. The number of supported sites is unlimited.

In addition, the login logic about `_NETRC_MACHINE` does not exist in the current `NiconicoChannelPlus*IE` extractors, so, in this PR, I didn't set `_NETRC_MACHINE = 'niconicochannelplus'` like #8645.

That is, keeping those `NiconicoChannelPlus*IE` subclasses is just for compatibility since they are in the current codebase. We could remove the "niconicochannelplus.py" file.

Reply:
`self._get_login_info(netrc_machine=self._DOMAIN)` is something to consider